File: done/141_list_cycle.py
##!/usr/bin/env python
'''
    @author [mst]
    @file   141_list_cycle.py    
    @brief  leetcode problems series

    based on the slightly harder 142 problem

    gains:
    -cycle algorithmic solving
    -sliding window technique

    features, changelog:
    -2022.12: -submission

    @version 0.1 2022.12
'''

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

# [demo] assisting function making a linked list out of a python list (array)
def make_list(nums):
    if not nums:
        return None
    
    head = ListNode(nums[0])
    p = head
    for n in nums[1::]:
        new_node = ListNode(n)
        p.next = new_node
        p = p.next

    return head

# [demo] assisting function making a linked list out of a python list (array)
def make_list_with_cycle(nums, pos):
    if not nums:
        return None
    
    head = ListNode(nums[0])
    p = head
    for n in nums[1::]:
        new_node = ListNode(n)
        p.next = new_node
        p = p.next

    if pos is not -1:
        cyc = head
        while pos is not 0:
            cyc = cyc.next
            pos = pos - 1
        p.next = cyc

    print ("made a list with cycle")

    return head


def print_list(lst : ListNode):
    while lst:
        print(f"{lst.val} -> ", end='')
        lst = lst.next
    print("")




# simple sliding-window solution:
# we first run a slow and a fast pointers to detect a cycle
# note: any None (nullptr) reference would mean an end of a list and so a lack of a cycle
# once the cycle node is spotted, we run a pointer out of a head of the list to spot the cycle entry
# 63% T 61 S
class Solution:
    def hasCycle(self, head: [ListNode]) -> bool:
        if head is None:
            return False
        
        slow = head
        fast = head

        while (slow is not None) and (fast is not None):
            if slow.next:
                slow = slow.next
            else:
                return False
            if fast.next and fast.next.next:
                fast = fast.next.next
            else:
                return False
            if slow == fast:
                return True
                
        return False



    def append(self, lst: ListNode, value) -> ListNode:
        if not lst:
            return ListNode(value)
        
        p = lst
        while p.next:
            p = p.next
        p.next = ListNode(value)

        return lst



################## DRIVER
def main():
    
    sol = Solution()
       
    # testing
    list1 = make_list_with_cycle([0,1,2,3], 1)
    # list2 = make_list([1,3,4])
    print(sol.hasCycle(list1))
    # print_list(list3)

    


if __name__ == ("__main__"):
    main()

File: done/141_list_cycle.cpp
/**
    @author [mst]
    @brief  leetcode problems series
	141_list_cycle
    Given head, the head of a linked list, determine if the
    linked list has a cycle in it.
    There is a cycle in a linked list if there is some node
    in the list that can be reached again by continuously following
    the next pointer. Internally, pos is used to denote the index of
    the node that tail's next pointer is connected to.
    Note that pos is not passed as a parameter.
    Return true if there is a cycle in the linked list. Otherwise, return false.

    gains:
    -cpp linked list basics

    @version 2023.02
*/


////////////////// LIBS
#include <iostream>            // usage of console prints
#include <vector>
#include <deque>
#include <algorithm>
using namespace std;

////////////////// DECL_IMPL

////////////////// DECL_IMPL

// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};


// print a list helper method
void printList (ListNode* head) {
    if (head == nullptr) {
        std::cout << "empty list" << std::endl;
        return;
    }

    ListNode* lp = head;
    while (lp) {
        std::cout << lp->val << ',';
        lp = lp->next;
    }

    std::cout << "nullptr" << '\n';
}

// list push helper function
void pushList(ListNode* &head, int val) {
    // [bp] this will still work with an empty list case
    ListNode* NewNode = new ListNode(val, head);
    head = NewNode;
}

// list end insertion helper function
void addTailList(ListNode* &head, int val) {
    //if (head == nulptr) { // [demo] unnecessary check
    //    return
    //}
    ListNode* NewNode = new ListNode(val, nullptr);
    if (head == nullptr) {
        head = NewNode;
        return;
    }

    ListNode* ptr = head;
    while (ptr->next) {
        ptr = ptr->next;
    }
    ptr->next = NewNode;
}

// detecting a cycle using the ususal slow and fast pointers approach
//
// sub: 38t 48s
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head) {
            ListNode* slow = head;
            ListNode* fast = head;

            while (slow && fast) {
                if (slow->next)
                    slow = slow->next;
                else
                    return false;

                if (fast->next && fast->next->next)
                    fast = fast->next->next;
                else
                    return false;

                if (slow == fast) return true;
            }
        }
        return false;
    }
};


////////////////// DRIVER
int main()
{
	Solution sol;

    // case 1
    ListNode* l5 = nullptr;
    addTailList (l5,3);
    addTailList (l5,2);
    addTailList (l5,0);
    addTailList (l5,-4);
    cout << "given list..." << endl;
    printList(l5);

    cout << "introducing a cycle and checking..." << endl;
    l5->next->next->next->next = l5->next;
    cout << sol.hasCycle(l5) << endl;

	return 0;
}
